dali/operators/image/resize/resize_base.cc

```diff
@@ -16,164 +16,117 @@
-  SubdivideInput(in_view);
-
-  out_shape_ = TensorListShape<>();
-  out_shape_.resize(in_view.num_samples(), in_view.sample_dim());
-  int sample_idx = 0;
-
-  kernels::KernelContext context;
-  context.gpu.stream = stream;
-  for (size_t b = 0; b < minibatches_.size(); b++) {
-    MiniBatch &mb = minibatches_[b];
-
-    auto &req = kmgr_.Setup<Kernel>(b, context,
-        mb.input, make_span(resample_params_.data() + mb.start, mb.count));
-
-    mb.out_shape = req.output_shapes[0];
-    // minbatches have uniform dim
-    for (int i = 0; i < mb.out_shape.size(); i++) {
-      out_shape_.set_tensor_shape(i + sample_idx, mb.out_shape[i]);
-    }
-    sample_idx += mb.out_shape.size();
-  }
 
-  output.Resize(out_shape_);
 
-  auto out_view = view<uint8_t, 3>(output);
-  SubdivideOutput(out_view);
-
-  for (size_t b = 0; b < minibatches_.size(); b++) {
-    MiniBatch &mb = minibatches_[b];
-
-    kmgr_.Run<Kernel>(0, b, context,
-        mb.output, mb.input, make_span(resample_params_.data() + mb.start, mb.count));
+template <>
+template <typename OutputType, typename InputType, int spatial_ndim>
+void ResizeBase<CPUBackend>::SetupResizeStatic(
+      TensorListShape<> &out_shape,
+      const TensorListShape<> &in_shape,
+      span<const kernels::ResamplingParams> params,
+      int first_spatial_dim) {
+  using ImplType = ResizeOpImplCPU<OutputType, InputType, spatial_ndim>;
+  auto *impl = dynamic_cast<ImplType*>(impl_.get());
+  if (!impl) {
+    impl_.reset();
+    auto unq_impl = std::make_unique<ImplType>(num_threads_);
+    impl = unq_impl.get();
+    impl_ = std::move(unq_impl);
   }
+  impl->Setup(out_shape, in_shape, first_spatial_dim, params);
 }
 
-void ResizeBase::Initialize(int num_threads) {
-  using Kernel = kernels::ResampleCPU<uint8_t, uint8_t>;
-  kmgr_.Resize<Kernel>(num_threads, num_threads);
-  out_shape_.resize(num_threads, 3);
-  resample_params_.resize(num_threads);
-}
 
-void ResizeBase::InitializeGPU(int batch_size, int mini_batch_size) {
-  DALI_ENFORCE(batch_size > 0, "Batch size must be positive");
-  DALI_ENFORCE(mini_batch_size > 0, "Mini-batch size must be positive");
-  const int num_minibatches = div_ceil(batch_size, mini_batch_size);
-  using Kernel = kernels::ResampleGPU<uint8_t, uint8_t>;
-  kmgr_.Resize<Kernel>(1, num_minibatches);
-  minibatches_.resize(num_minibatches);
-
-  for (int i = 0; i < num_minibatches; i++) {
-    int start = batch_size * i / num_minibatches;
-    int end = (batch_size * (i + 1)) / num_minibatches;
-
-    minibatches_[i].start = start;
-    minibatches_[i].count = end-start;
+template <>
+void ResizeBase<CPUBackend>::InitializeCPU(int num_threads) {
+  if (num_threads != num_threads_) {
+    impl_.reset();
+    num_threads_ = num_threads;
   }
+}
 
-  kmgr_.SetMemoryHint(kernels::AllocType::GPU, temp_buffer_hint_);
-  kmgr_.ReserveMaxScratchpad(0);
+template <>
+void ResizeBase<GPUBackend>::InitializeGPU(int minibatch_size) {
+  if (minibatch_size != minibatch_size_) {
```

Comment:
As above

Reply:
It's for lazy initialization - on first call, `minibatch_size_` is 0. We can call it in every SetupImpl and do not reallocate.